MF-596 - Add subtopic to RawMessage

beres opened the pull request:
### What does this do?

The patch add subtopic support to the following components
- cli
- http
- mqtt
- normalizer
- all readers
- sdk messages
- all writers
- ws

Not updated:
- coap
- lora

### Which issue(s) does this PR fix/relate to?
Put here `Resolves #XXX` to auto-close the issue that your PR fixes (if such)

### List any changes that modify/break current functionality

### Have you included tests for your changes?

### Did you document any new/modified functionality?

### Notes

beres committed:
Commit for mainflux-596
Modified and tested:
- cli
- http
- mqtt
- normalizer
- all readers
- sdk messages
- all writers
- ws
Missing:
- coap
- lora

Signed-off-by: ale <ale@metaverso.org>

chombium reviewed (changes requested):
Hi @beres, thanks for your contribution. The code looks good, but IMHO there are few things which need improvements:
1. As @drasko already mentioned on Gitter you should sign the commits.
2. There are few tests which are failing:
```
# github.com/mainflux/mainflux/readers/influxdb_test
readers/influxdb/messages_test.go:116:27: not enough arguments in call to reader.ReadAll
        have (string, uint64, uint64)
        want (string, uint64, uint64, map[string]string)
# github.com/mainflux/mainflux/readers/cassandra_test
readers/cassandra/messages_test.go:102:27: not enough arguments in call to reader.ReadAll
        have (string, uint64, uint64)
        want (string, uint64, uint64, map[string]string)
# github.com/mainflux/mainflux/readers/mongodb_test
readers/mongodb/messages_test.go:109:27: not enough arguments in call to reader.ReadAll
        have (string, uint64, uint64)
        want (string, uint64, uint64, map[string]string)
```
You can check if all the test are still passing with `make test`. Don't forget to start your docker daemon because the tests are using [Dockertest](https://github.com/ory/dockertest).
You can also run each test separately from the command line with for i.e. `GOCACHE=off go test -v -race readers/influxdb/messages_test.go`

3. There are no tests about the parts regarding the `subtopic` which you've added
4. Addition of few sentences about the `subtopics` in the user guide would also be nice

Keep up the good work.

Best Regards,
Jovan

mteodor commented on mqtt/mqtt.js:
```diff
@@ -85,14 +85,15 @@ function startMqtt() {
-        packet;
+        packet, subtopic;
     if (m && m.protocol !== 'mqtt') {
+        subtopic = m.subtopic !== '' ? m.subtopic.replace('.', '/') : ''
```
is the subtopic in .xxx.yyy.zzz format?

beres replied:
it must to be in xxx.yyy.zzz, fixed on last commit sorry

mteodor commented on mqtt/mqtt.js:
```diff
@@ -135,6 +136,7 @@ aedes.authorizePublish = function (client, packet, publish) {
                 rawMsg = message.RawMessage.encode({
                     publisher: client.thingId,
                     channel: channelId,
+                    subtopic: elements.join('.'),
```
or the subtopic is in xxx.yyy.zzz form

mteodor reviewed (commented):
overall looks ok

manuio commented on writers/cassandra/init.go:
```diff
@@ -12,6 +12,7 @@ import "github.com/gocql/gocql"
         id uuid,
         channel text,
+        subtopic text,
     	publisher text,
```
fix tabulation

dborovcanin commented on cli/message.go:
```diff
@@ -14,7 +14,7 @@ const contentTypeSenml = "application/senml+json"
 	cobra.Command{
 		Use:   "send",
-		Short: "send <channel_id> <JSON_string> <thing_key>",
+		Short: "send <channel_id>[.<topic>...] <JSON_string> <thing_key>",
```
Since we're using `subtopics` terminology, it's we should stay consistent and use `subtopics` or `subtopic` instead of `topic` in the cli.

dborovcanin commented on http/api/transport.go:
```diff
@@ -29,6 +31,9 @@ const protocol = "http"
 	errMalformedData = errors.New("malformed request data")
 	auth             mainflux.ThingsServiceClient
+	// regexp accept subtopic with name without '.' to avoid problems
+	//	with nats, / par of subtopic will be replaced with '.'
```
Fix indentation.

beres committed:
- fix subtopic name in, when starting with dot, http/ws/mqtt
- add some test on readers

Signed-off-by: ale <ale@metaverso.org>

dborovcanin commented on http/api/transport.go:
```diff
@@ -29,6 +31,9 @@ const protocol = "http"
 	auth             mainflux.ThingsServiceClient
+	// regexp accept subtopic with name without '.' to avoid problems
+	//	with nats, / par of subtopic will be replaced with '.'
+	channelPartRegExp = regexp.MustCompile(`^/channels/([\w\-]+)/messages(/[^?\.]+)*$`)
```
Maybe it's safer to allow only subtopics with alphanumeric characters, underscore, dash (`([\w\-])+`) and wildcard (`#`,  `+`, `$` or `*`). It will probably work this way, but I just don't like an idea of having any special character (e.g. `~`, `!` or  `%`) in subtopics.

drasko replied:
@dusanb94 @beres for this MQTT standard must be consulted. If it permits special chars in topics names then we must respect this.

beres replied:
Hi @dusanb94, yes it's safer, but why accept wildcard chars? http/api has only Publish method, what did you think?

dborovcanin replied:
Hi, @beres. I wrote this having MQTT and an idea of making subtopics as generic as possible on my mind. However, wildcards are related to subscriptions, so it should not be a problem here.

dborovcanin commented on http/nats/publisher.go:
```diff
@@ -33,6 +35,9 @@ func (pub *natsPublisher) Publish(msg mainflux.RawMessage) error {
 
-	subject := fmt.Sprintf("channel.%s", msg.Channel)
+	subject := fmt.Sprintf("%s.%s", prefix, msg.Channel)
+	if msg.Subtopic != "" {
```
You can use `fmt.Sprintf` here, as well.

dborovcanin commented on readers/cassandra/messages.go:
```diff
@@ -75,3 +73,31 @@ func (cr cassandraRepository) ReadAll(chanID string, offset, limit uint64) []mai
+			"publisher",
+			"name",
+			"protocol":
+			condSql += fmt.Sprintf(` AND %s = ?`, name)
```
You can use `fmt.Sprintf` to append: `condSql = fmt.Sprintf("%s AND %s = ?", condSql, name)`.

drasko replied:
Yes, @dusanb94 's proposition is more readable IMHO

dborovcanin commented on readers/influxdb/messages.go:
```diff
@@ -58,6 +59,26 @@ func (repo *influxRepository) ReadAll(chanID string, offset, limit uint64) []mai
+			"channel",
+			"subtopic",
+			"publisher":
+			condition += fmt.Sprintf(` AND %s='%s'`, name,
```
Same here and in the next `case`.

dborovcanin reviewed (changes requested):
Thanks for the contribution, @beres!

Please sign your commits, add the reference to an existing issue in PR description (#596), remove all `TODOs` from the code and check CI.

anovakovic01 commented on http/api/transport.go:
```diff
@@ -41,7 +46,7 @@ func MakeHandler(svc mainflux.MessagePublisher, tc mainflux.ThingsServiceClient)
 	r := bone.New()
 
-	r.Post("/channels/:id/messages", kithttp.NewServer(
+	r.Post("/channels/*", kithttp.NewServer(
```
Can we use `/channels/:id/*` here?

anovakovic01 commented on http/api/transport.go:
```diff
@@ -55,7 +60,19 @@ func MakeHandler(svc mainflux.MessagePublisher, tc mainflux.ThingsServiceClient)
+	}
+
+	chanID := channelParts[1]
+	subtopic := strings.Replace(channelParts[2], "/", ".", -1)
```
You can use `ReplaceAll` here.

beres committed:
- fix string concatenation
- update http/transport regexp to match subtopic names with only \w-
- update ws/transport regexp like http ones with also the wildcard * and >

Signed-off-by: ale <ale@metaverso.org>

anovakovic01 commented on mqtt/mqtt.js:
```diff
@@ -85,14 +85,16 @@ function startMqtt() {
-        packet;
+        packet, subtopic;
     if (m && m.protocol !== 'mqtt') {
+        subtopic = m.subtopic !== '' ? '/' + m.subtopic.replace('.', '/') : ''
```
You missed `;` at the end.

anovakovic01 commented on readers/cassandra/messages_test.go:
```diff
@@ -96,10 +99,24 @@ func TestReadAll(t *testing.T) {
+			andQuery: map[string]string{"subtopic": "not-present"},
+			messages: []mainflux.Message{},
+		},
+		fmt.Sprintf("read message with subtopic: %s", subtopic): {
```
Just specify if subtopic exists or not, not need for using `fmt.Sprintf` and specifying subtopic value.

anovakovic01 commented on readers/influxdb/messages_test.go:
```diff
@@ -110,10 +113,24 @@ func TestReadAll(t *testing.T) {
+			andQuery: map[string]string{"subtopic": "not-present"},
+			messages: []mainflux.Message{},
+		},
+		fmt.Sprintf("read message with subtopic: %s", subtopic): {
```
Same here.

anovakovic01 commented on readers/mocks/messages.go:
```diff
@@ -29,7 +29,7 @@ func NewMessageRepository(messages map[string][]mainflux.Message) readers.Messag
 }
 
-func (repo *messageRepositoryMock) ReadAll(chanID string, offset, limit uint64) []mainflux.Message {
+func (repo *messageRepositoryMock) ReadAll(chanID string, offset, limit uint64, andQuery map[string]string) []mainflux.Message {
```
I think that this `and` prefix doesn't improve readability of this code, so maybe we can call this one just `query`.

anovakovic01 commented on readers/mongodb/messages_test.go:
```diff
@@ -103,10 +106,24 @@ func TestReadAll(t *testing.T) {
+			andQuery: map[string]string{"subtopic": "not-present"},
+			messages: []mainflux.Message{},
+		},
+		fmt.Sprintf("read message with subtopic: %s", subtopic): {
```
Same here.

beres committed:
added subtopic support to coap adapter

Signed-off-by: ale <ale@metaverso.org>

anovakovic01 commented on ws/api/transport.go:
```diff
@@ -47,7 +52,7 @@ func MakeHandler(svc ws.Service, tc mainflux.ThingsServiceClient, l log.Logger)
 
 	mux := bone.New()
-	mux.GetFunc("/channels/:id/messages", handshake(svc))
+	mux.GetFunc("/channels/*", handshake(svc))
```
Can we use `/channels/:id/*` here?

drasko replied:
Maybe even something like `channels/:id/messages(/)?.*`

beres replied:
Yes :id parts can be used, but with the current implementation id and subtopic are managed on the same way by regexp, otherwise you get id from GetValue and subtopic form regexp, or we use regexp for both but ignore :id definition on router. What did you think?

anovakovic01 replied:
@beres I think that it would be more readable if we get id from `GetValue` and subtopic using regex.

beres replied:
what did you think about something like:
	mux.GetFunc("/channels/:id/messages", handshake(svc))
	mux.GetFunc("/channels/:id/messages/{:subtopic[a-zA-Z0-9/-_*>]+}", handshake(svc))

mteodor replied:
I dont think this will work, have you tried this?

beres replied:
@mteodor you are right, it does not actually work with bone mux :see_no_evil:

drasko replied:
@beres did you actually test this implementation thoroughly, or you plan to break 0.8 for huge number of companies using Mainflux :)

beres replied:
@drasko no, i believed  something like that could be possible, but as i replyed to @mteodor with bone does not work, while on coap mux it works. 

This seems to be the more readable solutions base on your preferences
mux.GetFunc("/channels/:id/messages", handshake(svc))
mux.GetFunc("/channels/:id/messages/*", handshake(svc))

drasko replied:
Committing to Mainflux at this mature stage is big responsability, especially when change is so big and touches core.

I suggest that you bring back WIP label to this PR until you test everything manually very throughly.

We are planning the 0.8 release with this change for next Friday, we will not have so much time to find potential regression.

drasko replied:
Just for the record, I really appreciate your work here, and you are helping us a lot with taking this issue.

beres replied:
@drasko I probably did not explain it well before, it was a proposal. If you liked it, I implemented it, otherwise not.
I have tested all the changes anyway it's right you take the time you need. :)

drasko replied:
@beres we like it. Just need to put it on the quality level of Mainflux.

drasko replied:
@beres this looks like unresolved. Is this working finally?

anovakovic01 commented on ws/nats/publisher.go:
```diff
@@ -45,18 +50,34 @@ func New(nc *broker.Conn) ws.Service {
 	return &natsPubSub{nc, cb}
 }
 
+func (pubsub *natsPubSub) getDestChannel(chanID, subtopic string) string {
```
When naming go methods, we don't use `get` prefix.

anovakovic01 commented on ws/nats/publisher.go:
```diff
@@ -25,7 +27,10 @@ const (
+var (
+	_               ws.Service = (*natsPubSub)(nil)
+	errInvalidTopic            = errors.New("invalid topic")
+)
```
We separate error vars from interface checking ones.

beres replied:
Sorry @anovakovic01 i don't understand, for  example ws/api/transport declare its own errors, which is the best place to define it?

anovakovic01 replied:
Sorry, I wasn't clear enough. What I meant by this is that you should separate vars into logical blocks. What I mean by this is that you first declar your errors and then this var that checks interface implementation:
```
var (
    err1 = ...
    err2 = ...
)

.
.
.

var _ ws.Service = (*natsPubSub)(nil)
```

drasko replied:
@beres is this one done?

beres replied:
Yes with the commit 638efec

beres committed:
- update replace functions with replaceall when needed
- renamed getDestChannel to fmtSubject
- update api/transport and ws/transport route to be more readable
- fix mqtt syntax
- renamed func andQuery to query as suggested by @anovakovic01
- have a nice we :)

Signed-off-by: ale <ale@metaverso.org>

beres committed:
- fix error declaration on ws/nat/publisher
- fix regexp added missing allowed chars - and _ on coap/api/transport
- fix subtopic clean suffix / if present on coap/api/transport
- improve regexp on http and ws /api/transport, now does not accept url that do not strictly match
- add some ws subtopic tests

Signed-off-by: ale <ale@metaverso.org>

beres committed:
fix invalid GOCACHE in go1.12, replaced by -count 1, see https://golang.org/doc/go1.10#test

Signed-off-by: ale <ale@metaverso.org>

beres committed:
- enabled wildcard chars on coap/api/transport
- allow use special chars on http and ws api/transport

Signed-off-by: ale <ale@metaverso.org>

beres committed:
- use strings.Replace() insted ReplaceAll()

Signed-off-by: ale <ale@metaverso.org>

drasko commented on coap/api/transport.go:
```diff
@@ -68,7 +69,9 @@ func MakeCOAPHandler(svc coap.Service, tc mainflux.ThingsServiceClient, l log.Lo
 	pingPeriod = pp
 	r := mux.NewRouter()
 	r.Handle("/channels/{id}/messages", gocoap.FuncHandler(receive(svc))).Methods(gocoap.POST)
+	r.Handle("/channels/{id}/messages/{subtopic:[/a-zA-Z0-9-_*>+#]+}", gocoap.FuncHandler(receive(svc))).Methods(gocoap.POST)
```
I still do not see why you cannot have subtopic name like `temp!%%%(99)/room?$&/~33`

I.e. I think you should enable all special chars allowed by MQTT protocol. For this you must open MQTT spec and take a look at topic rules: http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718106

```
4.7.3 Topic semantic and usage
The following rules apply to Topic Names and Topic Filters:

All Topic Names and Topic Filters MUST be at least one character long [MQTT-4.7.3-1]
Topic Names and Topic Filters are case sensitive
Topic Names and Topic Filters can include the space character
A leading or trailing ‘/’ creates a distinct Topic Name or Topic Filter
A Topic Name or Topic Filter consisting only of the ‘/’ character is valid 
Topic Names and Topic Filters MUST NOT include the null character (Unicode U+0000) [Unicode] [MQTT-4.7.3-2]
Topic Names and Topic Filters are UTF-8 encoded strings, they MUST NOT encode to more than 65535 bytes [MQTT-4.7.3-3]. See Section 1.5.3
```

When I read this: `Topic Names and Topic Filters are UTF-8 encoded strings` - this basically means any possible [UTF-8 character](https://en.wikipedia.org/wiki/List_of_Unicode_characters)

beres committed:
- allow every chars on subtopics
- fix replace error on mqtt

Signed-off-by: ale <ale@metaverso.org>

beres committed:
fix cassandra test

Signed-off-by: ale <ale@metaverso.org>

beres committed:
fix ws test with invalid subtopic

Signed-off-by: ale <ale@metaverso.org>

beres committed:
Merge master into mainflux-596

beres committed:
Merge branch 'master' into mainflux-596

anovakovic01 commented on coap/api/transport.go:
```diff
@@ -177,6 +189,14 @@ func observe(svc coap.Service, responses chan<- string) handler {
+			}
+			subtopic = strings.Replace(subtopic, "/", ".", -1)
+		}
+
```
Can you extract this code into a separate reusable function? It's duplicate of https://github.com/mainflux/mainflux/pull/642/files#diff-09d0397eb6e4e223a55f2af7f5ae75f0R145

beres replied:
Yes also in ws adapter, what is the best place to put it? no adapters include shared libraries

anovakovic01 replied:
For now just make function per adapter.

drasko replied:
@beres please take this remark into account.

anovakovic01 commented on http/api/transport.go:
```diff
@@ -27,35 +29,53 @@ import (
+		return nil, errMalformedData
+	}
+
+	chanID := channelParts[1]
```
Why don't you use `bone.GetValue` method?

beres replied:
no reason, a misprint of a previous implementation

anovakovic01 commented on http/api/transport.go:
```diff
@@ -27,35 +29,53 @@ import (
-	auth             mainflux.ThingsServiceClient
+	errMalformedData  = errors.New("malformed request data")
+	auth              mainflux.ThingsServiceClient
+	channelPartRegExp = regexp.MustCompile(`^/channels/([\w\-]+)/messages((/[^/]+)*)*(\?.*)?$`)
```
I think that this `((/[^/]+)*)*` can be simplified with `((/[^/]+)*)?`. Also, @drasko do we want to add support for query parameters in HTTP adapter? I'm not sure that we should allow in subtopic characters such as `.`, `>` and `*`, which are specific to NATS.

beres replied:
actually http does not use query parameters, i use the same regexp for coap/ws/http. If you do not want a specific regexp per adapter, it could be placed in a shared place

dborovcanin replied:
@beres If you want to do this, `topics.go` file in the project root is a good place for that.

beres replied:
@dusanb94 could also be the place for common subtopic function?

dborovcanin replied:
@beres If that function is shared and does not accept any adapter-specific arguments, then I'd say yes.

beres committed:
- improve regexp on http/ws api/transport
- minor changes

Signed-off-by: ale <ale@metaverso.org>

beres committed:
- add generic function parseSubtopic on ws/http adapters

Signed-off-by: ale <ale@metaverso.org>

beres committed:
Merge branch 'master' into mainflux-596

beres committed:
- add generic function fmtSubtopic on coap adapter

Signed-off-by: ale <ale@metaverso.org>

anovakovic01 reviewed (approved):
LGTM!

drasko reviewed (approved):
LGTM

manuio reviewed (approved):
LGTM
